core/src/main/java/io/temporal/samples/earlyreturn/TransactionActivitiesImpl.java

```diff
@@ -0,0 +1,75 @@
+/*
+ *  Copyright (c) 2020 Temporal Technologies, Inc. All Rights Reserved
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.samples.earlyreturn;
+
+import io.temporal.failure.ApplicationFailure;
+
+public class TransactionActivitiesImpl implements TransactionActivities {
+
+  @Override
+  public Transaction initTransaction(TransactionRequest txRequest) {
+    System.out.println("Initializing transaction");
+    sleep(500);
+    if (txRequest.getAmount() <= 0) {
+      System.out.println("Invalid amount: " + txRequest.getAmount());
+      throw ApplicationFailure.newNonRetryableFailure(
+          "Non-retryable Activity Failure: Invalid Amount", "InvalidAmount");
+    }
+    // Simulate transaction ID generation
+    String txId = "TXID" + String.format("%010d", (long) (Math.random() * 1_000_000_0000L));
+    Transaction tx =
+        new Transaction(
+            txId,
+            txRequest.getSourceAccount(),
+            txRequest.getTargetAccount(),
+            txRequest.getAmount());
+
+    sleep(500);
+    return tx;
+  }
+
+  @Override
+  public void cancelTransaction(TransactionRequest txRequest) {
```

Comment:
I think `cancelTransaction` should also take the `Transaction` class since, in practice, you would need the transaction ID to cancel it right?

Reply:
I'm diverging more and more from the original Go sample (which is fine).

Right now, I think of as initialization as something that results in the 'minting' of a Transaction ID. And a failure of initialization causes an activity failure which triggers a cancellation.

I can't think of a way of having
1. InitTransaction cause an activity failure (fundamental to @stephanos original sample) AND
2. Returning a transaction ID that would mean the creation of a Transaction object I can then use in cancelTransaction

Thinking blank-slate for a moment, and about the purpose of an "Early Return" type sample. @Quinn-With-Two-Ns -- What do you think you would want to see from a scenario where a transaction fails on initialization (e.g. the workflow update with start)? What data should be returned and what compensation should take place after? I don't think I should spend any more hours chewing on this sample without figuring that out first. Thanks!

Reply:
Yeah you raise a great point. I think I would separate the creation of the ID from the `initialization` of the transaction. Treating the transaction ID as an idempotency key for the rest of the activities. That being said I think this sample is really good as is and an improvement over the Go Sample so 
 wouldn't hold up the PR on this